In [13]:
# importing libraries
import pandas as pd
import tarfile
import warnings 
warnings.filterwarnings('ignore')

In [15]:
# Specify the path to your tar file
tar_path = r"C:\Users\Administrator\Downloads\yelp_dataset.tar"

# Open the tar file and save to a hidden file
with tarfile.open(tar_path, 'r') as tar:
    # Extract all contents to a specific directory
    tar.extractall(path=r"C:\Users\Administrator\Downloads\yelp_dataset")

## FIRST DATASET CONTAINING BUSINESS INFORMATION

In [16]:
# File Path to JSON file
file_path = r"C:\Users\Administrator\Downloads\yelp_dataset\yelp_academic_dataset_business.json"

# Load the JSON file into a DataFrame,
business = pd.read_json(file_path, lines=True)

# Displaying first 5 entries
business.head()

business_id                      name  \
0  Pns2l4eNsfO8kk83dixA6A  Abby Rappoport, LAC, CMQ   
1  mpf3x-BjTdTEA3yCZrAYPw             The UPS Store   
2  tUFrWirKiKi_TAnsVWINQQ                    Target   
3  MTSW4McQd7CbVtyjqoe9mw        St Honore Pastries   
4  mWMc6_wTdE0EUBKIGXDVfA  Perkiomen Valley Brewery   

                           address           city state postal_code  \
0           1616 Chapala St, Ste 2  Santa Barbara    CA       93101   
1  87 Grasso Plaza Shopping Center         Affton    MO       63123   
2             5255 E Broadway Blvd         Tucson    AZ       85711   
3                      935 Race St   Philadelphia    PA       19107   
4                    101 Walnut St     Green Lane    PA       18054   

    latitude   longitude  stars  review_count  is_open  \
0  34.426679 -119.711197    5.0             7        0   
1  38.551126  -90.335695    3.0            15        1   
2  32.223236 -110.880452    3.5            22        0   
3  39.955505  -75.155564    4.0            80        1   
4  40.338183  -75.471659    4.5            13        1   

                                          attributes  \
0                      {'ByAppointmentOnly': 'True'}   
1             {'BusinessAcceptsCreditCards': 'True'}   
2  {'BikeParking': 'True', 'BusinessAcceptsCredit...   
3  {'RestaurantsDelivery': 'False', 'OutdoorSeati...   
4  {'BusinessAcceptsCreditCards': 'True', 'Wheelc...   

                                          categories  \
0  Doctors, Traditional Chinese Medicine, Naturop...   
1  Shipping Centers, Local Services, Notaries, Ma...   
2  Department Stores, Shopping, Fashion, Home & G...   
3  Restaurants, Food, Bubble Tea, Coffee & Tea, B...   
4                          Brewpubs, Breweries, Food   

                                               hours  
0                                               None  
1  {'Monday': '0:0-0:0', 'Tuesday': '8:0-18:30', ...  
2  {'Monday': '8:0-22:0', 'Tuesday': '8:0-22:0', ...  
3  {'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', ...  
4  {'Wednesday': '14:0-22:0', 'Thursday': '16:0-2...

In [17]:
business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150346 entries, 0 to 150345
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   business_id   150346 non-null  object 
 1   name          150346 non-null  object 
 2   address       150346 non-null  object 
 3   city          150346 non-null  object 
 4   state         150346 non-null  object 
 5   postal_code   150346 non-null  object 
 6   latitude      150346 non-null  float64
 7   longitude     150346 non-null  float64
 8   stars         150346 non-null  float64
 9   review_count  150346 non-null  int64  
 10  is_open       150346 non-null  int64  
 11  attributes    136602 non-null  object 
 12  categories    150243 non-null  object 
 13  hours         127123 non-null  object 
dtypes: float64(3), int64(2), object(9)
memory usage: 16.1+ MB


In [18]:
# Filter the DataFrame to keep only rows that contain the word "restaurant"
restaurants_df = business[business['categories'].str.contains("restaurant", case=False, na=False)]

# Display the first few rows of the filtered DataFrame
restaurants_df.head()

business_id                   name              address  \
3   MTSW4McQd7CbVtyjqoe9mw     St Honore Pastries          935 Race St   
5   CF33F8-E6oudUQ46HnavjQ         Sonic Drive-In        615 S Main St   
8   k0hlBqXX-Bt0vf1op7Jr1w  Tsevi's Pub And Grill    8025 Mackenzie Rd   
9   bBDDEgkFA1Otx9Lfe7BZUQ         Sonic Drive-In  2312 Dickerson Pike   
11  eEOYSgkmpB90uNA7lDOMRA  Vietnamese Food Truck                        

            city state postal_code   latitude  longitude  stars  review_count  \
3   Philadelphia    PA       19107  39.955505 -75.155564    4.0            80   
5   Ashland City    TN       37015  36.269593 -87.058943    2.0             6   
8         Affton    MO       63123  38.565165 -90.321087    3.0            19   
9      Nashville    TN       37207  36.208102 -86.768170    1.5            10   
11     Tampa Bay    FL       33602  27.955269 -82.456320    4.0            10   

    is_open                                         attributes  \
3         1  {'RestaurantsDelivery': 'False', 'OutdoorSeati...   
5         1  {'BusinessParking': 'None', 'BusinessAcceptsCr...   
8         0  {'Caters': 'True', 'Alcohol': 'u'full_bar'', '...   
9         1  {'RestaurantsAttire': ''casual'', 'Restaurants...   
11        1  {'Alcohol': ''none'', 'OutdoorSeating': 'None'...   

                                           categories  \
3   Restaurants, Food, Bubble Tea, Coffee & Tea, B...   
5   Burgers, Fast Food, Sandwiches, Food, Ice Crea...   
8   Pubs, Restaurants, Italian, Bars, American (Tr...   
9   Ice Cream & Frozen Yogurt, Fast Food, Burgers,...   
11         Vietnamese, Food, Restaurants, Food Trucks   

                                                hours  
3   {'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', ...  
5   {'Monday': '0:0-0:0', 'Tuesday': '6:0-22:0', '...  
8                                                None  
9   {'Monday': '0:0-0:0', 'Tuesday': '6:0-21:0', '...  
11  {'Monday': '11:0-14:0', 'Tuesday': '11:0-14:0'...

In [19]:
restaurants_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 52286 entries, 3 to 150340
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   business_id   52286 non-null  object 
 1   name          52286 non-null  object 
 2   address       52286 non-null  object 
 3   city          52286 non-null  object 
 4   state         52286 non-null  object 
 5   postal_code   52286 non-null  object 
 6   latitude      52286 non-null  float64
 7   longitude     52286 non-null  float64
 8   stars         52286 non-null  float64
 9   review_count  52286 non-null  int64  
 10  is_open       52286 non-null  int64  
 11  attributes    51720 non-null  object 
 12  categories    52286 non-null  object 
 13  hours         45007 non-null  object 
dtypes: float64(3), int64(2), object(9)
memory usage: 6.0+ MB


In [24]:
# Saving restaurant data to CSV
restaurants_df.to_csv(r"C:\Users\Administrator\Downloads\yelp_dataset\restaurants.csv", index= False)


## SECOND DATASET CONTAINING REVIEWS INFORMATION

In [25]:
# File Path to  JSON file
file_path = r"C:\Users\Administrator\Downloads\yelp_dataset\yelp_academic_dataset_review.json"

# Define a chunk size
chunk_size = 10000

# Initialize an empty list to hold chunks
chunks = []

# Read the JSON file in chunks
with pd.read_json(file_path, lines=True, chunksize=chunk_size) as reader:
    for chunk in reader:
        # Process each chunk (e.g., filter, aggregate)
        chunks.append(chunk)

# Concatenate all chunks into a single DataFrame
reviews = pd.concat(chunks, ignore_index=True)

In [27]:
# Information of the reviews dataset
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6990280 entries, 0 to 6990279
Data columns (total 9 columns):
 #   Column       Dtype         
---  ------       -----         
 0   review_id    object        
 1   user_id      object        
 2   business_id  object        
 3   stars        int64         
 4   useful       int64         
 5   funny        int64         
 6   cool         int64         
 7   text         object        
 8   date         datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 480.0+ MB


In [28]:
# Getting the unique values in the restaurant_df's business Id column
restaurants_ids = restaurants_df['business_id'].unique()
len(restaurants_ids)

52286

In [29]:
# Filtering the reveiws dataframe to include only reviews related to restaurants
rest_reviews = reviews[reviews['business_id'].isin(restaurants_ids)]

In [30]:
# Getting resturant reviews infor
rest_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4724684 entries, 0 to 6990279
Data columns (total 9 columns):
 #   Column       Dtype         
---  ------       -----         
 0   review_id    object        
 1   user_id      object        
 2   business_id  object        
 3   stars        int64         
 4   useful       int64         
 5   funny        int64         
 6   cool         int64         
 7   text         object        
 8   date         datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 360.5+ MB


In [31]:
 # first 5 entries in restaurant reviews
rest_reviews.head()

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
2  saUsX_uimxRlCVr67Z4Jig  8g_iMtfSiwikVnbP2etR0A  YjUWPpI6HXG530lwP-fb2A   
3  AqPFMleE6RsU23_auESxiA  _7bHUi9Uuf5__HHc_Q8guQ  kxX2SOes4o-D3ZQBkiMRfA   
4  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   
5  JrIxlS1TzJ-iCu79ul40cQ  eUta8W_HdHMXPzLBBZhL1A  04UD14gamNjLY0IDYVhHJg   

   stars  useful  funny  cool  \
0      3       0      0     0   
2      3       0      0     0   
3      5       1      0     1   
4      4       1      0     1   
5      1       1      2     1   

                                                text                date  
0  If you decide to eat here, just be aware it is... 2018-07-07 22:09:11  
2  Family diner. Had the buffet. Eclectic assortm... 2014-02-05 20:30:30  
3  Wow!  Yummy, different,  delicious.   Our favo... 2015-01-04 00:01:03  
4  Cute interior and owner (?) gave us tour of up... 2017-01-14 20:54:15  
5  I am a long term frequent customer of this est... 2015-09-23 23:10:31

In [32]:
# Timeframe of the dataset
rest_reviews.date.min(), rest_reviews.date.max()

(Timestamp('2005-02-16 03:23:22'), Timestamp('2022-01-19 19:48:25'))

In [33]:
# Starting date of the dataframe to get 5 year worth of data
start_date = '2017-01-01 00:00:00'

# Getting 5 years worth of reviews
five_year_reviews = rest_reviews[(rest_reviews['date'] >= start_date)]

# first five entries
five_year_reviews.head()

review_id                 user_id             business_id  \
0   KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
4   Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   
24  lUUhg8ltDsUZ9h0xnwY4Dg  RreNy--tOmXMl1en0wiBOg  cPepkJeRMtHapc_b2Oe_dw   
30  -P5E9BYUaK7s3PwBF5oAyg  Jha0USGDMefGFRLik_xFQg  bMratNjTG5ZFEA6hVyr-xQ   
31  YbMyvlDA2W3Py5lTz8VK-A  4hBhtCSgoxkrFgHa4YAD-w  bbEXAEFr4RYHLlZ-HFssTA   

    stars  useful  funny  cool  \
0       3       0      0     0   
4       4       1      0     1   
24      4       1      0     1   
30      5       0      0     0   
31      5       0      0     0   

                                                 text                date  
0   If you decide to eat here, just be aware it is... 2018-07-07 22:09:11  
4   Cute interior and owner (?) gave us tour of up... 2017-01-14 20:54:15  
24  I was really between 3 and 4 stars for this on... 2018-07-17 03:30:07  
30  First time there and it was excellent!!! It fe... 2017-02-19 13:32:05  
31  Great burgers,fries and salad!  Burgers have a... 2017-01-02 03:17:34

In [34]:
# Getting information on the 5 year reviews dataset
five_year_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2559586 entries, 0 to 6990279
Data columns (total 9 columns):
 #   Column       Dtype         
---  ------       -----         
 0   review_id    object        
 1   user_id      object        
 2   business_id  object        
 3   stars        int64         
 4   useful       int64         
 5   funny        int64         
 6   cool         int64         
 7   text         object        
 8   date         datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 195.3+ MB


In [35]:
# Filtering to have only the relevant dataset
five_year_reviews=five_year_reviews[['user_id','business_id','stars','date']]

five_year_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2559586 entries, 0 to 6990279
Data columns (total 4 columns):
 #   Column       Dtype         
---  ------       -----         
 0   user_id      object        
 1   business_id  object        
 2   stars        int64         
 3   date         datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(2)
memory usage: 97.6+ MB


In [36]:
# saving the one_year_reviews to users.csv
five_year_reviews.to_csv(r"C:\Users\Administrator\Downloads\yelp_dataset\users.csv", index= False)

## SUMMARY

The JSON datasets used were filtered for ease of use as follows:

#### 1. Business Dataset
* The original dataset comprised 150,346 unique business entries, which were narrowed down to 52,286 entries after filtering to include only restaurants.
* The refined data was saved as restaurant.csv in the data folder.

#### 2. Reviews Dataset
* Initially, this dataset contained around 6.99 million reviews from various businesses.
* It was first filtered to include only restaurant reviews, reducing it to 4.7 million reviews.
* Further filtering was applied to focus on reviews from the past five years (2017-2022), bringing the count down to approximately 2.5 million reviews.
* This approach was used to maintain data relevancy and optimize computational efficiency.
The final dataset was saved as users.csv in the data folder.